common/model/bundle.c

```diff
@@ -51,3 +55,52 @@ void calculate_bundle_hash(bundle_transactions_t *bundle, flex_trit_t *out) {
   flex_trits_from_trits(out, NUM_TRITS_HASH, bundle_hash_trits, NUM_TRITS_HASH,
                         NUM_TRITS_HASH);
 }
+
+void bundle_finalize(bundle_transactions_t *bundle) {
+  iota_transaction_t *curr_tx = NULL;
+  bool valid_bundle = false;
+  iota_transaction_t *head_tx = NULL;
+  byte_t normalized_hash[HASH_LENGTH_TRYTE];
+  trit_t increased_tag_trits[NUM_TRITS_TAG];
+  flex_trit_t bundle_hash[FLEX_TRIT_SIZE_243];
+
+  Kerl kerl = {};
+  init_kerl(&kerl);
+
+  while (!valid_bundle) {
+  update_hash:
+    // bundle hash
+    bundle_calculate_hash(bundle, bundle_hash);
+    // normalize
+    normalize_hash(bundle_hash, normalized_hash);
+    kerl_reset(&kerl);
+    // checking 'M'
+    for (int i = 0; i < HASH_LENGTH_TRYTE; i++) {
+      if (normalized_hash[i] == 13) {
+        // Insecure bundle. Increment Tag and recompute bundle hash.
+        head_tx = (iota_transaction_t *)utarray_front(bundle);
+        flex_trits_to_trits(increased_tag_trits, NUM_TRITS_TAG,
+                            transaction_obsolete_tag(head_tx), NUM_TRITS_TAG,
+                            NUM_TRITS_TAG);
+        add_trits(one, increased_tag_trits, NUM_TRITS_TAG);
```

Comment:
I think you can do `add_assign(increased_tag_trits, NUM_TRITS_TAG, 1)`; hence no need for that `one` static array

Reply:
cool, confused about the name of `add_assign`, could you explain more?

Reply:
It modifies the trits you give it by operating a trit addition on it.
For example, it's used to obtain a subseed by adding an index to a seed: https://github.com/iotaledger/entangled/blob/develop/common/sign/v1/iss.c.inc#L22